pkg/m3admin/namespace/namespace.go

```diff
@@ -0,0 +1,218 @@
+func WithDomain(name string) Option {
+	return optionFn(func(n *namespace) error {
+		n.serviceDomain = name
+		return nil
+	})
+}
+
+// WithLogger is a setter to override the default logger
+func WithLogger(logger *zap.Logger) Option {
+	return optionFn(func(n *namespace) error {
+		n.logger = logger
+		return nil
+	})
+}
+
+// New constructs a new namespace client
+func New(opts ...Option) (Namespace, error) {
+	logging.InitWithCores(nil)
+	ctx := context.Background()
+	logger := logging.WithContext(ctx)
+	// TODO(PS) Add logger.Sync() somewhere
+	ns := &namespace{
+		client:        retryhttp.NewClient(),
+		logger:        logger,
+		servicePort:   m3admin.DefaultServicePort,
+		serviceName:   m3admin.DefaultServiceName,
+		serviceDomain: m3admin.DefaultServiceDomain,
+	}
+	for _, o := range opts {
+		if err := o.execute(ns); err != nil {
+			return nil, err
+		}
+	}
+	return ns, nil
+}
+
+// formatURL provides a formatted URL for HTTP reqeusts
+func (n *namespace) formatURL() string {
+	return fmt.Sprintf("http://%s.%s:%d", n.serviceName, n.serviceDomain, n.servicePort)
+}
+
+// defaultNamespaceReqeust provides sane defaults for a namespace reqeust
+func defaultNamespaceRequest(namespaceName string) *admin.NamespaceAddRequest {
+	return &admin.NamespaceAddRequest{
+		Name: namespaceName,
+		Options: &namespacepb.NamespaceOptions{
+			BootstrapEnabled:  _defaultBootstrapEnabled,
+			FlushEnabled:      _defaultFlushEnabled,
+			WritesToCommitLog: _defaultWritesToCommitLog,
+			CleanupEnabled:    _defaultCleanupEnabled,
+			RepairEnabled:     _defaultRepairEnabled,
+			RetentionOptions: &namespacepb.RetentionOptions{
+				RetentionPeriodNanos:                     _defaultRententionOptionRetentionPeriodNanos,
+				BlockSizeNanos:                           _defaultRententionOptionBlockSizeNanos,
+				BufferFutureNanos:                        _defaultRententionOptionBufferFutureNanos,
+				BufferPastNanos:                          _defaultRententionOptionBufferPastNanos,
+				BlockDataExpiry:                          _defaultRententionOptionBlockDataExpiry,
+				BlockDataExpiryAfterNotAccessPeriodNanos: _defaultRententionOptionBlockDataExpiryAfterNotAccessPeriodNanos,
+			},
+			SnapshotEnabled: _defaultSnapshotEnabled,
+			IndexOptions: &namespacepb.IndexOptions{
+				Enabled:        _defaultIndexOptionenabled,
+				BlockSizeNanos: _defaultIndexOptionBlockSizeNanos,
+			},
+		},
+	}
+}
+
+// Create will create a namespace
+func (n *namespace) Create(namespaceName string) error {
+	url := fmt.Sprintf("%s/%s", n.formatURL(), ns.AddURL)
+	data, err := json.Marshal(defaultNamespaceRequest(namespaceName))
+	if err != nil {
+		return err
+	}
+	_, err = m3admin.DoHTTPRequest(n.client, "POST", url, bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	n.logger.Info("successfully created namespace", zap.String("namespace", namespaceName))
+	return nil
+}
+
+// Get will retrieve a namespace
```

Comment:
which namespace will it retrieve? if it's all of them - then this it's probably better to call this `List()` 

also, could you use golang native types we've exposed - e.g. `[]string` for namespace names, or a slice of https://github.com/m3db/m3/blob/master/src/dbnode/storage/namespace/types.go#L108

Reply:
Updated it.